design/build-cache.md

```diff
@@ -0,0 +1,201 @@
+   1. Dependency caching for build systems such as npm, sbt and cargo.
+
+The build cache and cache usage is configurable via Yaml and backed by an internal sd-command and caching service.
+
+## Details
+
+### Yaml specification
+
+```yml
+shared:
+  # Caches defined in the shared section can be pipeline or exection scoped.
+  cache:
+    # A pipeline scoped cache.
+    pipeline:  
+      # The caching targets
+      target: ["build/**", "~/.sbt"]
+    # An event scoped cache
+    event:
+      target: ["target/deployment.jar"]
+    # A job scoped cache 
+    my-cache-1:
+      target: ["build/some-artifact.zip"]
+jobs:
+  test:
+    requires: [~pr]
+    image: rust:latest
+    steps:
+      - test: cargo test
+  build: 
+    requires: [~pr]
+    # Indicate which job scoped caches to use.
+    cache: ["my-cache-1"]
+    image: rust:latest
+    steps:
+      - test: cargo build
+  publish-preview:
+    # Indicate which job scoped caches to use.
+    cache: ["my-cache-1"]
+    requires: [~commit]
+    image: rust:latest
+    steps:
+      - add_musl: rustup target install x86_64-unknown-linux-musl
+      - build: cargo build --release --target=x86_64-unknown-linux-musl
+```
+
+### Design considerations
+
+1. Resource usage
+   1. Copying to and from a remote server for every job in a pipeline can lead to unnecessary network bandwidth usage and unnecessarily slow speeds.
+2. Integrity and consistency of cached data.
+3. User experience
+   1. Is it easily adaptable to new use cases?
+   2. Is it easily enabled and administered?
+   3. Is it easily configurable by end users once enabled?
+   4. Does it speed up build time? How much does it speed up build time for various use cases?
+   5. Is it consistent?
+   6. Is it reliable?
+4. Security
+   1. Is it possible to inject malicious code into pipelines via the cache mechanism? Malicious pull requests are a good example of this.
+
+### Security
+
+Injecting malicious code into a build via the caching mechanism is a real concern. This can lead to theft of tokens or credentials from job, or in even lead to malicious code being deployed to a production environment. 
+
+Pull requests, as the main vector for injection of malicious code, require additional security considerations. Pull requests therefore treat caches as readonly. 
+
+Caching of potentially sensitive data also needs to be handled. Data provided via the secrets field should not be cached. 
+
+### Cache invalidation
+
+- [Proposed] Changing environment variables or secrets should invalidate existing caches. Mass invalidation can be avoided by including a hash of variables and secrets with the cache metadata.
+- [Proposed] The user should also be able to invalidate caches via the UI.
+
+### Conflicts
+
+Multiple caches trying to restore the same file/files will result in an error. Caches by default do not overwrite existing files or directories. This behaviour can be configured by setting selecting a merging strategy.
```

Comment:
When will `Multiple caches trying to restore the same file/files` happen? Is this referring to multiple builds accessing the same cache directory at the same time?

Reply:
Hmm so I guess to clarify, we are talking about the event-level scope. And there are 2 edge cases:
1. The same job reads multiple cache directories, and there could be overlap between these directories
2. Multiple jobs can write back the same cache directory if they run in parallel

For the 1st case, I think it would make sense for us to simply say that we don't support the use case and it's user's responsibility to make sure that there is no overlap, otherwise the result could be unpredictable. We can implement validation for the user, but it can be of low priority.

The second case is actually a bit trickier IMO. If job A and B run in parallel and A finishes before B, we might lose what's added to the cache dir in job A. But I guess since this is just cache, the only impact should be build time.

Reply:
Yeah, the first case deals with writing to build space, the second case deals with writing to the cache.

## Case 1

> I think it would make sense for us to simply say that we don't support the use case and it's user's responsibility to make sure that there is no overlap, otherwise the result could be unpredictable.

Ditto. Leaving this as a userland responsibility is the best choice in my opinion, as well. 

> We can implement validation for the user, but it can be of low priority.

I thought about implementing a validation lint for the user, but it could get tricky if the user is using templates and an update in the templates causes case (1) to happen. 

It would still be a good idea to clearly define the order multiple caches get applied in. An on the spot suggestion would be to apply from broadest to narrowest scope. Multiple caches within the same scope get applied in order they are listed in (for job scopes) or defined in (for pipeline or event scopes).

## Case 2
> But I guess since this is just cache, the only impact should be build time.

Essentially, yes. The caching side of things needs to properly handle this case though. (Not really a problem using modern frameworks.)

## Conclusion

1. We leave case 1 as a user issue, but we do take care to explicitly define the order of cache application. The implementation should avoid parallelization of cache extraction as this can lead to increased unpredictability, unless special care is taken to ensure reproducibility.
2. The cache implementation must be able to handle multiple concurrent write requests for the same cache. The last write gets persisted.